projects/ad9361/src/ad9361_tinyiiod.c

```diff
@@ -0,0 +1,2721 @@
+{
+	ssize_t ret = 0;
+	u32 val = 0;
+	bool res = false;
+	ad9361_phy->pdata->fdd_independent_mode = false;
+
+	if (strequal(buf, "tx"))
+		val = ENSM_STATE_TX;
+	else if (strequal(buf, "rx"))
+		val = ENSM_STATE_RX;
+	else if (strequal(buf, "alert"))
+		val = ENSM_STATE_ALERT;
+	else if (strequal(buf, "fdd"))
+		val = ENSM_STATE_FDD;
+	else if (strequal(buf, "wait"))
+		val = ENSM_STATE_SLEEP_WAIT;
+	else if (strequal(buf, "sleep"))
+		val = ENSM_STATE_SLEEP;
+	else if (strequal(buf, "pinctrl")) {
+		res = true;
+		val = ENSM_STATE_SLEEP_WAIT;
+	} else if (strequal(buf, "pinctrl_fdd_indep")) {
+		val = ENSM_STATE_FDD;
+		ad9361_phy->pdata->fdd_independent_mode = true;
+	} else
+		return -ENOENT;
+
+	ad9361_set_ensm_mode(ad9361_phy, ad9361_phy->pdata->fdd, res);
+	ret = ad9361_ensm_set_state(ad9361_phy, val, res);
+
+	return ret;
+}
+
+/**
+ * set_multichip_sync
+ * @param *buff value to be written to attribute
+ * @param len of the value
+ * @param *channel channel properties
+ * @return length of chars written to attribute, or negative value on failure
+ */
+ssize_t set_multichip_sync(char *buf, size_t len,
+			   const struct channel_info *channel)
+{
+	ssize_t ret = 0;
+	uint32_t readin = read_ul_value(buf);
+	ret = ad9361_mcs(ad9361_phy, readin);
+
+	return ret;
+}
+
+/**
+ * set_filter_fir_config
+ * @param *buff value to be written to attribute
+ * @param len of the value
+ * @param *channel channel properties
+ * @return length of chars written to attribute, or negative value on failure
+ */
+ssize_t set_filter_fir_config(char *buf, size_t len,
+			      const struct channel_info *channel)
+{
+	return ad9361_parse_fir(ad9361_phy, (char *)buf, len);
+}
+
+static struct attrtibute_map global_write_attrtibute_map[] = {
+	{"trx_rate_governor", set_trx_rate_governor},
+	{"dcxo_tune_coarse", set_dcxo_tune_coarse},
+	{"dcxo_tune_fine", set_dcxo_tune_fine},
+	{"calib_mode", set_calib_mode},
+	{"ensm_mode", set_ensm_mode},
+	{"multichip_sync", set_multichip_sync},
+	{"filter_fir_config", set_filter_fir_config},
+};
+
+/**
+ * write global attribute
+ * @param *device name
+ * @param *attr name
+ * @param *buff where value is stored
+ * @param len length of the value
+ * @debug
+ * @return length of chars written to attribute, negative value in case of failure
+ */
+static ssize_t write_attr(const char *device, const char *attr,
+			  const char *buf, size_t len, bool debug)
+{
+	if (!dev_is_ad9361_module(device))
+		return -ENODEV;
+	if(strequal(device, "ad9361-phy")) {
+		int16_t attribute_id = get_attribute_id(attr, global_write_attrtibute_map,
+							ARRAY_SIZE(global_write_attrtibute_map));
+		if(attribute_id >= 0) {
+			return global_write_attrtibute_map[attribute_id].exec((char*)buf, len, NULL);
+		}
+		if(strequal(attr, "")) {
+			return write_all_attr((char*)buf, len, NULL, global_write_attrtibute_map,
+					      ARRAY_SIZE(global_write_attrtibute_map));
+		}
+		return -ENOENT;
+	} else if(strequal(device, "cf-ad9361-dds-core-lpc")) {
```

Comment:
double-check: should these be empty?

Reply:
Since the "cf-ad9361-dds-core-lpc" and "cf-ad9361-lpc" devices don't have any global attributes, yes, they should be empty. They only have channel attributes, that are treated in "ch_write_attr" and "ch_read_attr" functions.